src/recap.c

```diff
@@ -393,80 +395,241 @@ int process_pkt_read_state(struct pcap_stream_context *pctx) {
+    case PCAP_FILE_STATE_READ_PACKET:
       if (process_pkt_read_state(pctx) < 0) {
         log_error("process_pkt_read_state fail");
         return -1;
       }
       return 1;
-    case PCAP_STATE_FIN:
+    case PCAP_FILE_STATE_FIN:
       return 0;
     default:
       log_trace("Unknown state");
       return -1;
   }
 }
 
+int process_file_stream(const char *pcap_path, struct recap_context *pctx) {
+  if (pcap_path != NULL) {
+    if ((pctx->pcap_fd = fopen(pcap_path, "rb")) == NULL) {
+      log_errno("fopen failed for pcap file %s", pcap_path);
+      return -1;
+    }
+  } else {
+    pctx->pcap_fd = stdin;
+  }
+
+  int ret;
+  while ((ret = process_file_stream_state(pctx) > 0)) {
+  }
+
+  if (ret < 0) {
+    log_error("process_file_stream_state fail");
+    return -1;
+  }
+
+  return 0;
+}
+
+void add_packet_queue(UT_array *packets, struct packet_queue *queue) {
+  struct tuple_packet *p = NULL;
+
+  while ((p = (struct tuple_packet *)utarray_next(packets, p)) != NULL) {
+    if (push_packet_queue(queue, *p) == NULL) {
+      // Free the packet if cannot be added to the queue
+      free_packet_tuple(p);
+    }
+  }
+}
+
+void pcap_callback(const void *ctx, const void *pcap_ctx, char *ltype,
+                   struct pcap_pkthdr *header, uint8_t *packet) {
+
+  (void)pcap_ctx;
+
+  struct pcap_context *pc = (struct pcap_context *)pcap_ctx;
+  struct pcap_stat ps;
+
+  if (get_pcap_stats(pc, &ps) == 0) {
+    log_trace("ps_recv=%d ps_drop=%d ps_ifdrop=%d", ps.ps_recv, ps.ps_drop,
+              ps.ps_ifdrop);
+  }
+
+  struct recap_context *pctx = (struct recap_context *)ctx;
+
+  UT_array *packets = NULL;
+  utarray_new(packets, &tp_list_icd);
+
+  int npackets = extract_packets(ltype, header, packet, pctx->ifname, packets);
+
+  if (npackets > 0) {
+    add_packet_queue(packets, pctx->pq);
+    pctx->npackets += npackets;
+  }
+}
+
+void eloop_read_fd_handler(int sock, void *eloop_ctx, void *sock_ctx) {
+  (void)sock;
+  (void)sock_ctx;
+
+  struct pcap_context *pc = (struct pcap_context *)eloop_ctx;
+
+  if (capture_pcap_packet(pc) < 0) {
+    log_trace("capture_pcap_packet fail");
+  }
+}
+
+void save_packets_from_queue(struct recap_context *pctx) {
+  struct packet_queue *el = NULL;
+  while (is_packet_queue_empty(pctx->pq) < 1) {
+    if ((el = pop_packet_queue(pctx->pq)) != NULL) {
+      if (save_tuple_packet(pctx, &(el->tp)) < 0) {
+        log_error("save_tuple_packet fail");
+      }
+
+      free_packet_tuple(&el->tp);
+      free_packet_queue_el(el);
+    }
+  }
+}
+
+void eloop_tout_header_handler(void *eloop_ctx, void *user_ctx) {
+  struct recap_context *pctx = (struct recap_context *)user_ctx;
+
+  if (is_packet_queue_empty(pctx->pq) < 1) {
+    log_trace("Commiting packets to %s database", pctx->out_path);
+    if (execute_sqlite_query(pctx->db, "BEGIN IMMEDIATE TRANSACTION") < 0) {
+      log_error("Failed to capture a lock on db %s, please retry this "
+                "command later",
```

Comment:
You may need to improve this error. I'd even drop down this to a `log_warning()`, because I'd expect this to happen pretty often due to `SQLITE_BUSY`.

Reply:
Done